### PR TITLE
Enforce the HUB lock order

### LINKED ISSUE #66: Path to avoid deadlocks
Our `HUB` has a bunch of storages that we lock for read or write where needed. Once the entry points start being called from different threads, we can easily deadlock between two entry points. I think we can solve this by enforcing the order of locks: any entry point has to strictly lock the hub resources in the order their storages are declared (in `hub.rs`).

## Explanation
Supposing we make a node of a graph for every storage, and we mark an arrow from A to B to C if an entry point locks storage A followed by B followed by C. A deadlock occurs if at some point these arrows form a cycle, i.e. entry X locks A -> B -> C, while entry Y locks B -> C -> A. Of course, a cycle can also be formed by more than two entry points.

Now, let's assume the code follows a strict order for locking like suggested above. Imagine we hit a deadlock. Let's take a look at the last storage type being locked (pick a subset of all storages that are locked and take the last according to our order), call it A. Since we are in the deadlock, it's locked by some entry X and some other entries try (but fail) to lock it as well. However, there is nothing preventing X from proceeding, since whatever else it needs has to be locked *after* A, and those are not locked at this point. So X can proceed and free A eventually, so this is not a deadlock.

## Plan
  1. Audit the code and re-order the locks.
  2. Make sure that non-HUB resources are also locked in the order (whatever it is).
  3. Document the rules thoroughly in the code.
  4. Implement run-time debug checks for the order, asserting accordingly.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #66
cc @jrmuizel @m4b

### CODE REVIEW

**Comment by kvark** [comment]
This doesn't look great at all, unfortunately. Neither the implementation of `Token`, or the use of it in locking things. But it's the best thing we can do at the moment, and I think it's worth landing.
Also, I tested all the apps I have, this PR works for me.

File: wgpu-native/src/hub.rs [hunk 1]
```diff
@@ -136,9 +136,14 @@ impl<T, I: TypedId> Storage<T, I> {
 
 
 /// Type system for enforcing the lock order on shared HUB structures.
-/// The main property we are trying to establish is that each type
-/// of resources have only a single path from Root{} to reach by,
-/// when locking multiple resources types.
+/// If type A implements `Access<B>`, that means we are allowed to proceed
+/// with locking resource `B` after we lock `A`.
+///
+/// The impmenentations basically describe the edges in a directed graph
```
Comment (grovesNL) [inline wgpu-native/src/hub.rs]: typo: implementations

File: wgpu-native/src/device.rs [hunk 2]
```diff
@@ -253,14 +253,27 @@ impl<B: hal::Backend> PendingResources<B> {
...
 
+type DeviceToken<'a> = Token<'a, Device<back::Backend>>;
+
 impl PendingResources<back::Backend> {
-    fn triage_referenced(&mut self, trackers: &mut TrackerSet) {
+    fn triage_referenced(&mut self, trackers: &mut TrackerSet, mut token: &mut DeviceToken) {
+        // Before destruction, a resource is expected to have the following strong refs:
+        //  1. in resource itself
+        //  2. in the device tracker
+        //  3. in this list
```
Comment (grovesNL) [inline wgpu-native/src/device.rs]: I think we need to add a fourth item to this list?

Comment (kvark) [reply wgpu-native/src/device.rs]: actually, no :)
once it has 4 references, it's no longer a resource "before destruction"

**Comment by kvark** [comment]
@grovesNL fixed the typo and CI, PTAL!

**Comment by kvark** [comment]
@grovesNL rebased and resolved the conflict. Would you want to finish this review?

**Pull request merged**